### PR TITLE
Fix SSL certificate cache refresh and collision handling.

### PR DESCRIPTION
**Pull request by chtsanti**
SslBump was ignoring some origin server certificate changes or differences,
incorrectly using the previously cached fake certificate (mimicking 
now-stale properties or properties of a slightly different certificate).
Also, Squid was not detecting key collisions inside certificate caches.

On-disk certificate cache fixes:

  - Use the original certificate signature instead of the certificate
    subject as part of the key. Using signatures reduces certificate key
    collisions to deliberate attacks and woefully misconfigured origins,
    and makes any mishandled attacks a lot less dangerous because the
    attacking origin server certificate cannot by trusted by a properly
    configured Squid and cannot be used for encryption by an attacker.

    We have considered using certificate digests instead of signatures.
    Digests would further reduce the attack surface to copies of public
    certificates (as if the origin server was woefully misconfigured).
    However, unlike the origin-supplied signatures, digests require
    (expensive) computation in Squid, and implemented collision handling
    should make any signature-based attacks unappealing. Signatures won
    on performance grounds.

    Other key components remain the same: NotValidAfter, NotValidBefore,
    forced common name, non-default signing algorithm, and signing hash.

  - Store the original server certificate in the cache (together with
    the generated certificate) for reliable key collision detection.

  - Upon detecting key collisions, ignore and replace the existing cache
    entry with a freshly computed one. This change is required to
    prevent an attacker from tricking Squid into hitting a cached
    impersonating certificate when talking to a legitimate origin.

In-memory SSL context cache fixes:

  - Use the original server certificate (in ASN.1 form) as a part of the
    cache key, to completely eliminate cache key collisions.

Other related improvements:

  - Make the LruMap keys template parameters.
  - Polish Ssl::CertificateDb class member names to match Squid coding
    style. Rename some functions parameters to better match their
    meaning.
  - Replace Ssl::CertificateProperties::dbKey() with:
     * Ssl::OnDiskCertificateDbKey() in ssl/gadgets.cc for
       on-disk key generation by the ssl_crtd helper;
     * Ssl::InRamCertificateDbKey() in ssl/support.cc for
       in-memory binary keys generation by the SSL context memory cache.
  - Optimization: Added Ssl::BIO_new_SBuf(SBuf*) for OpenSSL to write
    directly into SBuf objects.

This is a Measurement Factory project.

### CODE REVIEW

**Comment by rousskov** [comment]
@chtsanti, it looks like our basic CI testing found at least two problems. Clicking on "Details" under "All checks have failed" and then expanding the failed tests all the way through (by clicking on the pinkish names of failed test components) will eventually get you to the failed test log inside Semaphore CI. The "git diff --check" test is a part of the PR [Checklist](http://wiki.squid-cache.org/MergeProcedure#Submission_Checklist) that you can easily reproduce on your machine. See [GitHints](http://wiki.squid-cache.org/GitHints#Compare_your_changes_with_the_official_code).

After you have fixed build and formatting problems, please [squash](http://wiki.squid-cache.org/GitHints#Squash_all_the_feature_branch_changes_into_a_single_commit) and [sync](http://wiki.squid-cache.org/GitHints#Rebase_your_feature_branch_to_be_in_sync_with_the_current_upstream_master) your changes so that the PR can be merged "as is".

**Comment by chtsanti** [comment]
Unfortunately I did not follow Alex notes.

I merged the  SQUID-210-Certificate-Cache-refresh-fixes  with latest squid/master so the "squash" and "sync" procedure can not be followed (by the way is it normal to follow such procedure? We may have very complex branches, developed for long time and this is will not work).

Should I create new PR?

Also there is one more error in semaphoreci, the X509_get0_signature function not found.

**Comment by rousskov** [comment]
> I merged the SQUID-210-Certificate-Cache-refresh-fixes with latest squid/master so the "squash" and "sync" procedure can not be followed

Why not?! The procedure should work regardless of your branch state. You may have to be more careful with selecting the right fork point, resolve an extra conflict, and/or undo a useless commit before squashing, but _anything_ can be squashed and/or rebased. Remember that you are no trying to preserve your branch history in this case. You are cleaning it up to end up with a single commit.

> Should I create new PR?

I see no point in doing that. It is your branch that matters, not the PR ID.

> Also there is one more error in semaphoreci, the X509_get0_signature function not found.

Please clarify: Do you consider that error to be a Semaphore CI problem? Or our/Squid problem? I assume that it is the latter, that this PR introduced this problem, and that your branch will eventually fix it, but please correct me if I am wrong.

**Comment by rousskov** [comment]
> by the way is it normal to follow such procedure?

It is 50% "normal". Many well-run projects either require squashed PRs or will squash the changes automatically. This is the "normal" part. Rebasing/syncing should be done automatically in most simple cases, but we do not have the automation for that (yet). 

The exact steps are up to each developer, of course. The steps outlined on GitHints is just one way to accomplish a single-commit, synced feature branch.

> We may have very complex branches, developed for long time and this is will not work.

The branch history or complexity does not matter at all! If you can produce a patch file, then I can give you a synced branch that contains a single commit.  And that is all we want. The GitHints steps accomplish the same result with (safer) git commands, but those steps are not important. It is the final result (a single-commit branch) that is important.

```
git checkout -b topic-clean upstream/master
patch -p0 < /tmp/my-patch-from-a-long-lived-topic-branch.patch
git add ...
git commit
git push ...
hub pull-request
```

File: src/base/LruMap.h [hunk 1]
```diff
@@ -10,42 +10,43 @@
 #define SQUID_LRUMAP_H
 
 #include "SquidTime.h"
+#include "sbuf/SBuf.h"
```
Comment (chtsanti) [inline src/base/LruMap.h]: fixed

**Comment by chtsanti** [comment]
> It is 50% "normal". Many well-run projects either require squashed PRs or will squash the changes automatically. This is the "normal" part. Rebasing/syncing should be done automatically in most simple cases, but we do not have the automation for that (yet).

OK, I wanted to ask why not just use "git merge --squash" option, I suppose we are going to use something similar when the related scripts be ready.

But still, according github documents, there is the option "merge and squash" to allow merge the PR branch as one commit.  Isn't enough?

```
git checkout -b topic-clean upstream/master
patch -p0 < /tmp/my-patch-from-a-long-lived-topic-branch.patch
git add ...
git commit
git push ...
hub pull-request
```
but it is possible that the sequel is:
```
git commit
git merge upstream/master
git commit
git merge upstream/master
git merge upstream/master
```
So your patch commits are mixed with merge commits. This is what I did into this branch.
Even if someone did not do any merge from upstream/master but just do a rebase after the final review on squashed branch, it is possible that new commits/fixes are required.

Anyway, I think there are two options for this branch now:
  - Make new PR.
  - Do a "git reset (--hard)" on this branch, apply the final patch and follow the GitHints.

**Comment by rousskov** [comment]
> I wanted to ask why not just use "git merge --squash" option

A successful `git merge --squash upstream/master` will place missing upstream/master commits into your branch as as single commit, on _top_ of your branch changes. The end result (tip of the code) will be the same as rebase+squash but, AFAICT, it can make future merging very difficult. Also, if somebody tries to review your individual commits, they are more likely to be confused because those commits will show changes to the old master code.

`git merge` is more likely to fail and/or produce difficult-to-resolve conflicts, especially when the master ran away pretty far. It essentially applies each master commit to all of your feature changes (as opposed to applying a single squashed feature commit to the top of the master branch).

There are cases where `git merge` would work better, but, on average, I expect it to be worse.

> But still, according github documents, there is the option "merge and squash" to allow merge the PR branch as one commit. Isn't enough?

It is enough to get your code into the master. However, 

1. The button requires the committer to guess whether the author wanted to squash all the changes or actually cherished every commit. 
1. Squashing may destroy authorship information (again without an explicit author permission).
1. It will maim the commit message, requiring the committer to do submitter's job of editing/polishing it.
1. When this button is available, squashing is easy and the author can do it in seconds.

There are other serious problems with Green Buttons related to CI tests. Eventually, commits will be automated to avoid those problems. That, in turn, will put even more pressure to get your branch in the right order before it is merged.

On the bright side, we should also automate conflict-free squashing and rebasing so that developers can simply request it from the bot -- almost like clicking half of the Green Button.

> but it is possible that the sequel is: ...

It does not matter what the sequence was: You can _always_ generate a diff/patch from the end result. Yes, that patch may not apply to master "as is", but if that is the case, then there is work for you to do no matter what -- the magic Green Button will not be available because of the conflicts.

> if someone did not do any merge from upstream/master but just do a rebase after the final review on squashed branch, it is possible that new commits/fixes are required.

Yes, of course. There is no way around that, with or without the Green Button. There is no magic that can resolve all conflicts automatically. Needless to say, a reviewer may refuse to start the review until the code is reasonably up to date, may not approve the code until it is up to date,  and/or may request additional changes after the "new commits/fixes".

> Make new PR.

A new PR changes nothing in this case. It is your feature branch state that matters. A PR is just an (official branch, feature branch) pair with comments/votes.

> Do a "git reset (--hard)" on this branch, apply the final patch and follow the GitHints.

I do not think you want to apply the final patch (against the current master) to the result of "git reset" which, I presume, will move you back to the old master. I would expect something like this:

```
git checkout -b SQUID-210-Certificate-Cache-refresh-fixes-clean upstream/master # new branch
patch -p0 < /tmp/SQUID-210-vs-current-master.patch # no conflicts if your patch is correct
git add ... # add any new files (or you could try "git am" instead of "patch" above?)
git commit # a single commit; all changes are squashed
git push --force origin:SQUID-210-Certificate-Cache-refresh-fixes # GitHub will notice changes and update this PR
```


Disclaimer: I am not a git expert. There are most likely ways to improve my sketches or even completely replace them with some other, better ones.

**Comment by yadij** [comment]
Perhapse if you think of it this way:
* your branch is based on the master branch, so when master changes you need to rebase (eventually).
* master is not based on your code so to combine them is a merge.

**Comment by chtsanti** [comment]
Please reopen, looks that I have not the rights to do it.

**Comment by chtsanti** [comment]
> git checkout -b SQUID-210-Certificate-Cache-refresh-fixes-clean upstream/master # new branch
> patch -p0 < /tmp/SQUID-210-vs-current-master.patch # no conflicts if your patch is correct
> git add ... # add any new files (or you could try "git am" instead of "patch" above?)
> git commit # a single commit; all changes are squashed
> git push --force origin:SQUID-210-Certificate-Cache-refresh-fixes

The correct git commands to replace a branch located at Organization/ABranch  are the followings:
```
git clone  git@github.com:Organization/squid.git  ABranch-clean
git checkout -b ABranch-clean
git add ...
git commit ...
git push --force origin ABranch-clean:ABranch
```
Unfortunately playing with this commands I removed the SQUID-210-Certificate-Cache-refresh-fixes branch and the PR is closed.

**Comment by yadij** [comment]
Maybe a new PR for the cleaned branch would be easier all round. This one is getting quite cluttered with extra details.

**Comment by chtsanti** [comment]
Maybe, it is easier.
However maybe it is better to continue into this PR until all problems solved and use the results to make git guides, or make procedures for other similar cases.
And finally after problems solved I will make a new PR for final commit.

The problems did not finished for me. I have rebased the branch locally but I can not update the measurement-factory:SQUID-210-Certificate-Cache-refresh-fixes, because the measurement-factory/squid.git is not updated.

**Comment by rousskov** [comment]
> The correct git commands to replace a branch located at Organization/ABranch are the followings:

These are incorrect. You are branching off stale origin/master. You should be branching off fresh upstream/master. See [What happens to origin/master?](https://wiki.squid-cache.org/GitHints#What_happens_to_origin.2Fmaster.3F)

My commands may have minor typos (git approach to naming targets is notoriously inconsistent so I often have to try different separators), but they should work in principle AFAICT. They do assume the same initial setup as in GitHints. YMMV.

**Comment by rousskov** [comment]
AFAICT, we cannot (easily) reopen this PR because @chtsanti has deleted the corresponding branch (and restoring that branch is not enough to convince GitHub to let us repopen this PR). See https://github.com/isaacs/github/issues/361

There is probably a [workaround](https://github.com/isaacs/github/issues/361#issuecomment-219088644), but it is so complex, that I hesitate suggesting it at this point. Besides, @chtsanti may not have the old HEAD saved to use that workaround anyway.

@chtsanti, please open a new PR.

**Comment by rousskov** [comment]
Polished commands, replacing `:` with a space in `git push`:
```
git checkout -b SQUID-210-Certificate-Cache-refresh-fixes-clean upstream/master # new branch
patch -p0 < /tmp/SQUID-210-vs-current-master.patch # no conflicts if your patch is correct
git add ... # add any new files (or you could try "git am" instead of "patch" above?)
git commit # a single commit; all changes are squashed
git push --force origin SQUID-210-Certificate-Cache-refresh-fixes # GitHub will notice changes and updates the PR linked to SQUID-210-Certificate-Cache-refresh-fixes
```

Needless to say, if origin/SQUID-210-Certificate-Cache-refresh-fixes does not exist, you do not need to force a push.

**Comment by rousskov** [comment]
Based on private Christos feedback, I suspect that when pushing from (local) branch A to (remote) branch B, git needs us to name both branches: `A:B`. Or, to be more precise, git needs us to name both

* what we want to push (i.e., the `<src>` part of push's `<revspec>`) _and_ 
* where we want to push it (i.e., the `<dst>` part of push `<revspec>`).

The `<refspec>` section of `git push --help` explains this reasonably well.

I have not tested this, but I suspect that `HEAD` can be used as `<src>` in this case:

```
git push --force origin HEAD:SQUID-210-Certificate-Cache-refresh-fixes # GitHub will notice...
```

Here is the complete recipe with the above change, for the record:

```
git checkout -b SQUID-210-Certificate-Cache-refresh-fixes-clean upstream/master # new branch
patch -p0 < /tmp/SQUID-210-vs-current-master.patch # no conflicts if your patch is correct
git add ... # add any new files (or you could try "git am" instead of "patch" above?)
git commit # a single commit; all changes are squashed
git push --force origin HEAD:SQUID-210-Certificate-Cache-refresh-fixes # GitHub will notice changes and updates the PR linked to SQUID-210-Certificate-Cache-refresh-fixes
```

**Pull request closed without merging**